torch/csrc/autograd/profiler_python.cpp

```diff
@@ -0,0 +1,707 @@
+        //  `track_module`, and `reset` is responsible for  calling Py_DECREF
+        //  when clearing `module_calls_`.
+        PyObject* self_;
+    };
+
+    bool active_;
+    PyObject* module_call_code_;
+    std::string path_prefixes_;
+    std::vector<TraceContext*> trace_contexts_;
+
+    std::vector<RawEvent> events_;
+    std::vector<ModuleForward> module_calls_;
+
+    using DescriptionKey = std::pair</*f_code=*/PyCodeObject*, /*f_lasti=*/int>;
+    ska::flat_hash_map<DescriptionKey, CodeDescription, hash_pair> code_descriptions_;
+};
+
+PythonTracer& PythonTracer::singleton() {
+    static PythonTracer singleton_;
+    return singleton_;
+}
+
+PythonTracer::PythonTracer() : active_(false) {
+    path_prefixes_ = py::module::import("torch.profiler.python_tracer")
+        .attr("_prefix_regex")().cast<std::string>();
+
+    module_call_code_ = py::module::import("torch.nn")
+        .attr("Module")
+        .attr("__call__")
+        .attr("__code__")
+        .ptr();
+}
+
+void PythonTracer::start(size_t max_threads) {
+    TORCH_CHECK(!active_, "PythonTracer is already active")
+    TORCH_CHECK(!trace_contexts_.size(), "PythonTracer should not have active contexts");
+    TORCH_CHECK(max_threads > 0, "max_threads must be positive, got ", max_threads);
+    TORCH_CHECK(
+        max_threads <= max_py_threads,
+        "max_threads must be less than or equal to ", max_py_threads);
+
+    pybind11::gil_scoped_acquire gil;
+    auto t0 = now();
+
+    // Loop over all interpreters and all threads within each interpreter.
+    // We will need to register a trace function with each thread. We set the
+    // current thread to position zero to ensure that it is traced, and so we
+    // can restore the thread state after registration.
+    std::vector<PyThreadState*> thread_states { PyThreadState_Get() };
+    if (max_threads > 1) {
+        PyInterpreterState* interpreter_state = PyInterpreterState_Head();
+        while (interpreter_state != nullptr) {
+            PyThreadState* thread_state = PyInterpreterState_ThreadHead(interpreter_state);
+            while (thread_state != nullptr) {
+                if (thread_state != thread_states[0]) {
+                    thread_states.push_back(thread_state);
+                }
+                thread_state = PyThreadState_Next(thread_state);
+            }
+            interpreter_state = PyInterpreterState_Next(interpreter_state);
+        }
+
+        if (thread_states.size() > max_threads) {
+            std::cout << "Warning: can only trace " << max_threads << " threads. "
+                    << thread_states.size() << " are currently active." << std::endl;
+            thread_states.resize(max_threads);
+        }
+    }
+
+    // Register the tracer in each thread.
+    for (size_t i = 0; i < thread_states.size(); i++) {
+        PyThreadState* thread_state = thread_states[i];
+        PyThreadState_Swap(thread_state);
```

Comment:
just curious why we need to restore the Python thread state to pre - starting the trace. seems like best practice by CPython and wondering if there is any good explanation on this?

Reply:
There isn't any particular reason why it's imperative, but mucking with the interpreter and then not cleaning it up just seemed like bad form.